fix(cli): handle symlinks in updater bundler, closes #3933

dceddia opened the pull request:
<!--
Update "[ ]" to "[x]" to check a box

Please make sure to read the Pull Request Guidelines: https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

### What kind of change does this PR introduce?
<!-- Check at least one. If you are introducing a new binding, you must reference an issue where this binding has been proposed, discussed and approved by the maintainers. -->

- [x] Bugfix
- [ ] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?
<!-- If yes, please describe the impact and migration path for existing applications in an attached issue. -->

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [x] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [ ] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [ ] I have added a convincing reason for adding this feature, if necessary

### Other information

dceddia committed:
fix(cli): handle symlinks in updater bundler

lucasfernog commented:
Nice catch! Though linking frameworks shows how slow the extract process is. I'll address that in another PR.

lucasfernog commented:
Do we also need to handle symlinks [here](https://github.com/tauri-apps/tauri/blob/c54ddfe9338e7eb90b4d5b02dfde687d432d5bc1/core/tauri/src/api/file/extract.rs#L220)? I'm not sure how macOS links frameworks in the app bundle.

dceddia commented:
Ooh yeah, probably need to handle symlinks at extraction time too. I imagine otherwise it’ll end up creating hard directories and files and/or falling over.

lucasfernog commented:
The #3963 change is essential for frameworks in the updater bundle since the extract process was super slow (and a framework folder has a lot of files).

dceddia committed:
fix(cli): correctly extract symlinks from the updater's .tar.gz

dceddia commented:
I added a test to check that symlinks inside frameworks extract correctly, and updated the `extract` function to handle symlinks. I noticed the built-in `entry.unpack()` function seems to handle all the different types of files, so I went with that. (it didn't handle directories unfortunately, so I left the branch that deals with those).

dceddia committed:
remove unused set_perms functions

dceddia commented:
Even though the test is passing and the symlinks look good after a test run, this isn't working in my actual app. Not sure why yet...

Update: oof, nevermind me. I had built and installed a local copy of `tauri-cli` but my app was still building against the rc10 from crates.io instead of this version with the fixes 🤦‍♂️ Once I point my Cargo.toml at this copy it works fine.

dceddia committed:
Merge branch 'dev' into copy-symlinks

dceddia commented:
Heads up that I think I found an issue with this where it's making symlinks to absolute paths, maybe only sometimes, maybe only if the app name has a space in it. I'm looking into it, but just wanted to leave a note here not to merge this yet.

dceddia committed:
fix(cli): avoid broken symlinks in updater bundle

The bundler was creating invalid symlinks that would show up with no
permissions when extracted with tar, which was breaking the updater in
some cases.

This changes the tar creation to use tar::Builder's append_dir_all which
handles recursively adding files, directories, and symlinks.

dceddia commented:
I think this is good now. I've been running with the last fix ("avoid broken symlinks in updater bundle") for a few releases of my app, haven't seen any problems myself, and haven't had any complaints from users about Mac updates being broken.

HeavenVolkoff commented:
Hello,

Just stumbled upon the issue linked to this PR while working on https://github.com/spacedriveapp/spacedrive/pull/768. Our app requires a `framework` to be bundled, but after linking it in `tauri.macOS.frameworks` the bundle step for the updater started to fail with the error message `Failed to tar.gz update directory: 'Is a directory (os error 21)'`.

Is there anything else that needs to be addressed for this PR to move forward?

lucasfernog commented:
Let's try to release it on 1.4, I'll let the auditors know about this.

lucasfernog committed:
Merge branch 'dev' into copy-symlinks

lucasfernog committed:
use current impl on linux

wusyong commented:
@lucasfernog what's breaking changes in this PR?

lucasfernog commented:
There's no breaking changes, it could be a patch too.

HeavenVolkoff commented:
> There's no breaking changes, it could be a patch too.

It would be great if this could be included in a patch release.

Also, the failed test seems to be just a typo: `bundle_path` should be `bundle_paths` here https://github.com/tauri-apps/tauri/pull/3934/files#diff-e7c0c60e9bef4a36a5940e86d4b58739b653f64ce5b116a84a6bd3a64f7268aaR317

lucasfernog committed:
Merge remote-tracking branch 'origin/dev' into copy-symlinks

lucasfernog committed:
update test

chippers commented on core/tauri/src/api/file/extract.rs:
```diff
@@ -100,13 +101,8 @@ impl<'a, R: Read> Entry<'a, R> {
-            set_perms(into_path, Some(&mut out_file), mode, true)?;
-          }
+          // handle files, symlinks, hard links, etc. and set permissions
+          entry.unpack(into_path)?;
```
Using [`Entry::unpack_in`](https://docs.rs/tar/latest/tar/struct.Entry.html#method.unpack_in) will help prevent unpacking a file outside the destination.
```suggestion
          entry.unpack_in(into_path)?;
```

lucasfernog replied:
We actually need to use unpack() here. The implementation is completely messed up, but it basically needs to extract `YourApp-0.1.1.app/Contents/MacOS/file` into `/tmp/tauri-update-blabla/ContentsMacOS/file` and later we rename `/tmp/tauri-update-blabla` to `/path/to/YourApp.app`

chippers replied:
oof

lucasfernog replied:
But what we can do it check if the path has any parent component and error out... I don't think it should ever happen.

lucasfernog committed:
fix test

lucasfernog committed:
Update core/tauri/src/api/file/extract.rs

Co-authored-by: chip <chip@chip.sh>

lucasfernog committed:
Revert "Update core/tauri/src/api/file/extract.rs"

This reverts commit 07c9d42e60436cd91946ed7b70f1a8b04d6ceb46.

lucasfernog committed:
Revert "fix extract"

This reverts commit e2d3e90199db3e0855d60ffb116365918c1f87b6.

lucasfernog committed:
fix extract

lucasfernog committed:
chore: ensure path does not contain parent dir component

lucasfernog committed:
fix impl

morajabi commented:
can we have this in v2 as well?

lucasfernog commented:
Yeah @morajabi we'll merge 1.x into dev for the last time when 1.5 lands.
